torch/_inductor/lowering.py

```diff
@@ -1567,6 +1567,21 @@ def apply_constraint(arg, fx_arg):
 make_fallback(aten.exponential.default, warn=False)
 
 
+@register_lowering(aten.copy)
+def copy(self, src, non_blocking=False):
+    x = src
+    if self.get_device() != src.get_device():
+        x = to_device(x, self.get_device())
+    if self.get_dtype() == src.get_dtype():
+        x = to_dtype(x, self.get_dtype())
+
+    # TODO: respect storage offset?
+    if self.get_size() != src.get_size():
+        out = expand(src, self.get_size())
+        return clone(out)
+    return clone(src)
```

Comment:
cc @ngimel does this lowering for `copy()` look reasonable to you?

Reply:
There's one test that's not passing: `test/inductor/test_torchinductor_dynamic_shapes.py DynamicShapesCpuTests.test_input_mutation2_dynamic_shapes_cpu`

```
        def fn(a):
            b = a + 1
            a.view(64).copy_(torch.tensor([66.0], device=a.device))
            c = a + 2
            return b, c
```

It fails with this:
```
AttributeError: 'Pointwise' object has no attribute 'get_stride'
```

And if I print the graph that inductor is trying to lower, I see:
```
def forward(self, primals_1):
    clone = torch.ops.aten.clone.default(primals_1)
    add = torch.ops.aten.add.Tensor(clone, 1)
    view = torch.ops.aten.view.default(clone, [64])
    _tensor_constant0 = self._tensor_constant0
    lift_fresh_copy = torch.ops.aten.lift_fresh_copy.default(_tensor_constant0);  _tensor_constant0 = None
    copy = torch.ops.aten.copy.default(view, lift_fresh_copy);  view = lift_fresh_copy = None
    sym_size = torch.ops.aten.sym_size(primals_1, 1);  primals_1 = None
    view_1 = torch.ops.aten.view.default(copy, [1, sym_size]);  copy = None
    add_1 = torch.ops.aten.add.Tensor(view_1, 2)
    sym_stride = torch.ops.aten.sym_stride(clone, 0);  clone = None
    return [view_1, add, add_1, sym_size, sym_stride]
```

Should we not expect to see a `sym_stride()` call in the inductor graph? I can dig through the test some more to see where it's coming from.

Reply:
I think this is some preexisting problem; in fact, hf_LongFormer is failing on dynamic shapes on master for exactly the same reason

Reply:
Hm I agree it's a preexisting problem, I haven't seen `sym_stride` in the inductor graph previously, perhaps we should register lowering for it.

Reply:
Upd: we already have a lowering for `sym_stride`, the problem is that `clone` is not a realized buffer, and thus `strides` for it are not defined. Should we instead return strides from the fx graph meta data?

Reply:
Getting the result from meta makes the test pass, add 
```
            elif n.target == torch.ops.aten.sym_stride:
                if isinstance(n.meta["val"], torch.SymInt):
                    result = n.meta["val"].node.expr
                else:
                    result = super().run_node(n)
```
in `run_node` in graph.py

Reply:
@bdhirsh Can you post the full backtrace here?

Reply:
Hmm, looks like we are saving the sym_stride so that we can use it in the backward (via an `empty_strided()` call):
```
def forward(self, sym_size, sym_stride, tangents_1, tangents_2, tangents_3):
    add_2 = torch.ops.aten.add.Tensor(tangents_1, tangents_3);  tangents_1 = tangents_3 = None
    new_empty_strided = torch.ops.aten.new_empty_strided.default(add_2, [1, sym_size], [sym_stride, 1]);  sym_stride = None
    copy_1 = torch.ops.aten.copy.default(new_empty_strided, add_2);  new_empty_strided = add_2 = None
    view_5 = torch.ops.aten.view.default(copy_1, [64]);  copy_1 = None
    clone_1 = torch.ops.aten.clone.default(view_5, memory_format = torch.contiguous_format)
    full_like = torch.ops.aten.full_like.default(clone_1, 0, pin_memory = False, memory_format = torch.contiguous_format);  clone_1 = None
    copy_2 = torch.ops.aten.copy.default(view_5, full_like);  view_5 = full_like = None
    view_6 = torch.ops.aten.view.default(copy_2, [1, sym_size]);  copy_2 = sym_size = None
    add_3 = torch.ops.aten.add.Tensor(view_6, tangents_2);  view_6 = tangents_2 = None
    return [add_3]
```

It looks like that's coming from autograd's `CopySlices`: https://github.com/pytorch/pytorch/blob/01476465ddb17824d00c9f5dbc39daf984f2267c/torch/csrc/autograd/functions/tensor.cpp#L79

Reply:
It seems like we're stuck with the possibility of saving strides for the backwards then? Unless we want to special case autograd's CopyBackwards decomposition for inductor.

Although even if we end up passing incorrect strides into the backward graph, does that really matter in practice? (since inductor is stride agnostic). That would mean the only remaining problem is "can we return output strides from an inductor forward graph without inductor complaining".

Reply:
It should be fine for us to propagate strides from forward to backward. Inductor can take the strides and ignore it (because it doesn't need to know about strides, only what the final striding is).